cql/src/main/java/org/apache/cassandra/stargate/metrics/ClientMetrics.java

```diff
@@ -88,62 +109,69 @@ public void unpauseConnection() {
     pausedConnections.decrementAndGet();
   }
 
-  public void markRequestDiscarded() {
-    requestDiscarded.mark();
-  }
-
-  public Counter getTotalBytesRead() {
-    return totalBytesRead;
+  public void incrementTotalBytesRead(double value) {
+    totalBytesRead.increment(value);
   }
 
-  public Counter getTotalBytesWritten() {
-    return totalBytesWritten;
+  public void incrementTotalBytesWritten(double value) {
+    totalBytesWritten.increment(value);
   }
 
-  public Histogram getBytesReceivedPerFrame() {
-    return bytesReceivedPerFrame;
+  public void recordBytesReceivedPerFrame(double value) {
+    bytesReceivedPerFrame.record(value);
   }
 
-  public Histogram getBytesTransmittedPerFrame() {
-    return bytesTransmittedPerFrame;
+  public void recordBytesTransmittedPerFrame(double value) {
+    bytesTransmittedPerFrame.record(value);
   }
 
-  public List<ConnectedClient> allConnectedClients() {
-    List<ConnectedClient> clients = new ArrayList<>();
-
-    for (Server server : servers) clients.addAll(server.getConnectedClients());
-
-    return clients;
-  }
+  public synchronized void init(
+      Collection<Server> servers,
+      MetricRegistry metricRegistry,
+      MeterRegistry meterRegistry,
+      ClientInfoMetricsTagProvider clientInfoTagProvider) {
 
-  public synchronized void init(Collection<Server> servers, MetricRegistry metricRegistry) {
     if (initialized) return;
 
     this.servers = servers;
     this.metricRegistry = metricRegistry;
+    this.meterRegistry = meterRegistry;
+    this.clientInfoTagProvider = clientInfoTagProvider;
+
+    // netty gauges
+    meterRegistry.gauge(
+        metric("NettyDirectMemory"), Tags.empty(), this, ClientMetrics::nettyDirectMemory);
+    meterRegistry.gauge(
+        metric("NettyHeapMemory"), Tags.empty(), this, ClientMetrics::nettyHeapMemory);
+
+    // connected native clients with the client info tags must be a multi gauge
+    connectedNativeClients =
+        MultiGauge.builder(metric("connectedNativeClients")).register(meterRegistry);
+
+    // TODO can we avoid using the timer, but get notified from somebody on the changes?
+    Timer connectedNativeClientsTimer = new Timer();
+    connectedNativeClientsTimer.schedule(
+        new TimerTask() {
+          @Override
+          public void run() {
+            updateConnectedClients();
+          }
+        },
+        10000L,
+        10000L);
```

Comment:
> can we avoid using the timer, but get notified from somebody on the changes?

Yeah a timer has a few drawbacks: holds a thread, updates only happen at a fixed interval...

And `Server` does not provide a way to monitor when client connections get added/removed. And even if we added that, I'm not sure if that's the best approach because we only really need to update the values when the metrics get scraped.

Maybe we could extend `PrometheusMeterRegistry` to add a `beforeScrape` callback?

Reply:
No matter if we take this proposal, **we need to fix the connection tracking removing anyway as @mpenick suggested above**.. This is such a nice way to have memory leaks..

Reply:
OK confirmed that there is no need to remove channels & connections, the `DefaultChannelGroup` does this automatically..